runtime/parachains/src/parathread_assigner.rs

```diff
@@ -0,0 +1,211 @@
+pub mod pallet {
+	use super::*;
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	// TODO Using libp2p `PeerId` or it's public key implementation instead of `CollatorId` could be
+	// beneficial for node side execution/filtering of peers.
+	type ScheduledCollator = CollatorId;
+
+	type BalanceOf<T> =
+		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The runtime's definition of an event.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The runtime's definition of a Currency.
+		type Currency: ReservableCurrency<Self::AccountId>;
+
+		/// An origin which is allowed to set the base price of parathreads.
+		type SetPriceOrigin: EnsureOrigin<<Self as frame_system::Config>::RuntimeOrigin>;
+
+		/// The maximum number of orders stored by the price controller. This should be bounded by the
+		/// ~atoms in the universe~ some sensible number
+		#[pallet::constant]
+		type MaxOrders: Get<u32>;
+	}
+
+	#[derive(Encode, Decode, Debug, TypeInfo, MaxEncodedLen, Eq, PartialEq, Clone)]
+	pub struct CoreOrder {
+		exotic_id: ParaId,
+		scheduled_collator: ScheduledCollator,
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_base_spot_price)]
+	pub(super) type BaseSpotPrice<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+	// The order storage. A simple bounded vec will do for now but this can be changed to a
+	// message queue at a later point.
+	#[pallet::storage]
+	pub type OrderBook<T> =
+		StorageValue<_, BoundedVec<CoreOrder, <T as Config>::MaxOrders>, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		BaseSpotPriceSet { amount: BalanceOf<T> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a single parathread core order.
+		///
+		/// Parameters:
+		/// - `origin`:
+		///
+		/// Errors:
+		/// Events:
+		/// TODO weights
+		#[pallet::call_index(0)]
+		#[pallet::weight(1_000)]
+		pub fn place_order(origin: OriginFor<T>) -> DispatchResult {
+			let _res = ensure_signed(origin);
+			Ok(())
+		}
+
+		/// Set the base spot price. Should only be available to the runtime's SetPriceOrigin.
+		///
+		/// Parameters:
+		/// - `origin`: Must be root
+		/// - `amount`: The price to set as the base
+		///
+		/// Errors:
+		/// - `BadOrigin`
+		/// Events:
+		/// - `BaseSpotPriceSet(amount)`
+		/// TODO weights
+		#[pallet::call_index(1)]
+		#[pallet::weight(1_000)]
+		pub fn set_base_spot_price(origin: OriginFor<T>, amount: BalanceOf<T>) -> DispatchResult {
+			// Ensure the proper origin authority
+			T::SetPriceOrigin::ensure_origin(origin)?;
+
+			BaseSpotPrice::<T>::set(amount);
+			Self::deposit_event(Event::BaseSpotPriceSet { amount });
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
```

Comment:
Shouldn't this be the implementation of [Hooks](https://github.com/paritytech/substrate/blob/b3d9f3c57e7f799d49442e3456614431f2d0e51a/frame/examples/basic/src/lib.rs#L387) right away?

Reply:
My initial intuition was to just use the hooks immediately like you mention. I switched to the initializer pattern because all of the pallets within the parachains module were using it.

I don't think that anything in this pallet is dependant on initalization/finalization order so it might be okay to just use hooks directly?

Reply:
Whatever makes sense, I was just curious.